File: .omo/re-galaxy/ExportGalaxy.py
# Ghidra Jython post-analysis export script
# Dumps: functions.jsonl, data-symbols.tsv, and xrefs to candidate galaxy strings/constmsg ids
import json
from ghidra.program.model.symbol import RefType

fm = currentProgram.getFunctionManager()
listing = currentProgram.getListing()
memory = currentProgram.getMemory()
refmgr = currentProgram.getReferenceManager()
symtab = currentProgram.getSymbolTable()
out_dir = r"E:\logh7-revival\.omo\re-galaxy"

# 1) functions dump
with open(out_dir + r"\functions.jsonl", "w") as f:
    for fn in fm.getFunctions(True):
        e = fn.getEntryPoint()
        body = fn.getBody()
        f.write(json.dumps({
            "name": fn.getName(),
            "entry": "0x%x" % e.getOffset(),
            "size": body.getNumAddresses(),
        }) + "\n")

# 2) defined data symbols with values (candidate tables)
with open(out_dir + r"\data-symbols.tsv", "w") as f:
    di = listing.getDefinedData(True)
    for d in di:
        addr = d.getAddress()
        dt = d.getDataType().getName()
        blk = memory.getBlock(addr)
        blkname = blk.getName() if blk else "?"
        try:
            val = d.getDefaultValueRepresentation()
        except:
            val = ""
        f.write("0x%x\t%s\t%s\t%s\n" % (addr.getOffset(), blkname, dt, val[:80]))

# 3) find string references to galaxy/system/planet keywords
keywords = ["galaxy","system","planet","star","Null_galaxy","seikei","wakusei",
            "grid","MAP","map_","cluster","sector","coord"]
hits = []
di = currentProgram.getListing().getDefinedData(True)
for d in di:
    try:
        v = d.getValue()
    except:
        continue
    if v is None: continue
    s = str(v)
    low = s.lower()
    for kw in keywords:
        if kw.lower() in low:
            addr = d.getAddress()
            refs = refmgr.getReferencesTo(addr)
            reflist = []
            for r in refs:
                reflist.append("0x%x" % r.getFromAddress().getOffset())
            hits.append({"addr":"0x%x"%addr.getOffset(),"str":s[:100],"refs":reflist})
            break
with open(out_dir + r"\string-hits.json","w") as f:
    json.dump(hits, f, indent=1)

print("EXPORT DONE: %d string hits" % len(hits))

File: .omo/re-galaxy/ExportGalaxy.java
// Java GhidraScript: locate galaxy/planet loaders + KV parser, dump xrefs & decompiled callers
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.data.StringDataInstance;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.app.decompiler.*;
import java.io.*;
import java.util.*;

public class ExportGalaxy extends GhidraScript {
    public void run() throws Exception {
        String OUT = "E:\\logh7-revival\\.omo\\re-galaxy\\";
        Listing listing = currentProgram.getListing();
        ReferenceManager rm = currentProgram.getReferenceManager();
        FunctionManager fm = currentProgram.getFunctionManager();

        // target strings to locate (case-insensitive substring match on defined data)
        String[] targets = {"galaxy.mdx","null_galaxy","p%03d_low.mdx","fs%03d_low.mdx",
            "fixedstar=","planet_num=","class_=","revolution_radius=","habitability=",
            "area=","grid_index=","start_year="};
        Map<String,List<Long>> strAddr = new LinkedHashMap<>();
        for (String t: targets) strAddr.put(t, new ArrayList<Long>());

        DataIterator di = listing.getDefinedData(true);
        while (di.hasNext()) {
            Data d = di.next();
            Object v = d.getValue();
            if (v == null) continue;
            String s = v.toString();
            for (String t: targets) {
                if (s.equals(t) || s.contains(t)) strAddr.get(t).add(d.getAddress().getOffset());
            }
        }

        // collect referencing functions
        DecompInterface dec = new DecompInterface();
        dec.openProgram(currentProgram);
        PrintWriter xr = new PrintWriter(new FileWriter(OUT+"galaxy-xrefs.txt"));
        Set<Function> toDecomp = new LinkedHashSet<>();
        for (String t: targets) {
            xr.println("### STRING "+t+" -> "+strAddr.get(t).size()+" instance(s)");
            for (long off: strAddr.get(t)) {
                Address a = toAddr(off);
                xr.println("  data @ 0x"+Long.toHexString(off));
                ReferenceIterator ri = rm.getReferencesTo(a);
                while (ri.hasNext()) {
                    Reference r = ri.next();
                    Address from = r.getFromAddress();
                    Function f = fm.getFunctionContaining(from);
                    xr.println("     xref from 0x"+Long.toHexString(from.getOffset())+
                        (f!=null?("  in "+f.getName()+" @0x"+Long.toHexString(f.getEntryPoint().getOffset())):"  (no func)"));
                    if (f!=null) toDecomp.add(f);
                }
            }
        }
        xr.close();

        PrintWriter dc = new PrintWriter(new FileWriter(OUT+"galaxy-decomp.c"));
        int n=0;
        for (Function f: toDecomp) {
            if (n++>40) { dc.println("// ...truncated at 40 funcs"); break; }
            DecompileResults res = dec.decompileFunction(f, 60, monitor);
            dc.println("// ===== "+f.getName()+" @0x"+Long.toHexString(f.getEntryPoint().getOffset())+" =====");
            if (res!=null && res.decompileCompleted())
                dc.println(res.getDecompiledFunction().getC());
            else dc.println("// decompile failed");
        }
        dc.close();

        // dump function list
        PrintWriter fl = new PrintWriter(new FileWriter(OUT+"functions.tsv"));
        FunctionIterator fi = fm.getFunctions(true);
        int cnt=0;
        while (fi.hasNext()) { Function f=fi.next(); fl.println("0x"+Long.toHexString(f.getEntryPoint().getOffset())+"\t"+f.getName()+"\t"+f.getBody().getNumAddresses()); cnt++; }
        fl.close();
        println("EXPORT DONE funcs="+cnt+" decomp="+Math.min(n,41)+" xref-funcs="+toDecomp.size());
    }
}
